### ISSUE TITLE
Diagnostic suppressed when bad code in macro

### ISSUE THREAD

**Issue by pogo59** [llvm/llvm-project#73022]
https://godbolt.org/z/zzMrTzv6a

We're seeing a warning diagnostic (reference cannot be null, so it always evaluates true) from plain code (e.g. line 6 below) but not when the identical code is produced from a macro (line 5).  This seems... wrong.
```
#define TEST(COND) if (COND) goo()
void goo(void);
void test(int &);
void test(int &a) {
  TEST(&a);
  if (&a) goo();
}
```

**Comment by hazohelet** [comment]
Clang suppresses some of its diagnostics when they originate from macros because macro expansion results may sometimes look incorrect to the compiler while macro authors are confident that they know what they are doing.
This case looks intentional as well. https://github.com/llvm/llvm-project/blob/331111277a3e80aedd1a6877524feadfbcb41a88/clang/lib/Sema/SemaChecking.cpp#L16142-L16148

**Comment by pogo59** [comment]
I have to say, it seems inconsistent to see:
```
TEST(&a); // not reported
TEST(&a == nullptr); // reported
```
I understand that the suppression is deliberate, but I suggest that this is misguided. I would like to understand a specific example where suppressing the warning in the former case is a good thing.

**Comment by dwblaikie** [comment]
A fairly simple example might be to imagine a macro that takes a pointer:
```
#define COND_CALL_FOO(x) x && x->foo()
```
And if you call `COND_CALL_FOO(&y)` - you might get a warning about a trivially true test, despite it not being trivially true once you combine them.
